src/com/ansorgit/plugins/bash/lang/lexer/bash.flex

```diff
@@ -168,27 +168,34 @@ Filedescriptor = "&" {IntegerLiteral} | "&-"
 %xstate X_HEREDOC_MARKER
 %xstate X_HEREDOC_MARKER_IGNORE_TABS
 %xstate X_HEREDOC
+// this state is used to keep the lexer in a non-zero state when heredoc markers were found
+// if used as a highlighting lexer incremental lexing is used when in a zero state
+// heredoc lexing is stateful and is thus not compatible with the zero state
+%state S_HEREDOC_EXPECTED
 
 /* To match here-strings */
 %xstate X_HERE_STRING
 
 %%
 /***************************** INITIAL STAATE ************************************/
-<YYINITIAL, S_CASE, S_CASE_PATTERN, S_SUBSHELL, S_ASSIGNMENT_LIST> {
+<YYINITIAL, S_CASE, S_CASE_PATTERN, S_SUBSHELL, S_ASSIGNMENT_LIST, S_HEREDOC_EXPECTED> {
   {Shebang}                     { return SHEBANG; }
   {Comment}                     { return COMMENT; }
 }
 
 <X_HEREDOC_MARKER, X_HEREDOC_MARKER_IGNORE_TABS> {
+    // a line-feed after a heredoc marker indicates a missing start marker tag,
+    // the lexer must not remain in the heredoc marker state
+    {LineTerminator}             { heredocState().removeMarker(zzCurrentPos); backToPreviousState(); return LINE_FEED; }
     {WhiteSpaceLineCont}+        { return WHITESPACE; }
     {LineContinuation}+          { return WHITESPACE; }
-    {LineTerminator}             { return LINE_FEED; }
 
       ("$"? "'" [^\']+ "'")+
     | ("$"? \" [^\"]+ \")+
     | [^ \t\n\r\f;&|]+ {
-        heredocState().pushMarker(yytext(), yystate() == X_HEREDOC_MARKER_IGNORE_TABS);
+        heredocState().pushMarker(zzCurrentPos, yytext(), yystate() == X_HEREDOC_MARKER_IGNORE_TABS);
         backToPreviousState();
```

Comment:
I think we can go without this action, in all cases on stack you will have state which was before `X_HEREDOC_MARKER` or `X_HEREDOC_MARKER_IGNORE_TABS`. Why we need to pop it and push it back again by  `goToState(S_HEREDOC_EXPECTED);`
I think both this rows can be replaced by single `yybegin(S_HEREDOC_EXPECTED);`

Reply:
I see.
We still need this to supported deeper nesting, I think.
For example:
```bash
echo $(
cat <<EOF
   content
EOF
)
```